typing/typemod.ml

```diff
@@ -1163,6 +1167,103 @@ let rec approx_modtype env smty =
       let aliasable = (not (Env.is_functor_arg path env)) in
       Mty_strengthen (mty, path, Aliasability.aliasable aliasable)
 
+and approx_with ~loc env sg constr =
+  (* We must remove the destructively substituted item so it doesn't falsely
+     shadow items in the signature with the same name. In the case of module
+     substitution, we also actually perform the substitution, to avoid a
+     dangling module alias.
+
+     Because the approximated signature resulting from this function is only
+     used to resolve names in other modules in the recursive groups, we need not
+     replace dangling references to types: types can only be referenced on the
+     RHS of a type declaration, but approximated modules treat all types as
+     abstract, so there is nothing to replace.
+
+     See [testsuite/tests/typing-recmod/regression_destructive_subst.ml] for
+     examples.
+
+     CR-someday: it would be much more robust to parameterize [transl_*]
+     functions by whether or not we are approximating, so they share the same
+     logic, as it is a common source of bugs for these functions to miss some
+     behavior. *)
+  let open struct
+    type namespace = Type | Module | Module_type
+  end in
+  let rec remove_from_sg namespace path env sg =
+    let replaced_path, filtered_sg =
+      List.fold_left_map
+        (fun replaced_path item ->
+           let replaced_path', item =
+             remove_from_sig_item namespace path env sg item
+           in
+           let replaced_path =
+             match replaced_path, replaced_path' with
+             | None, None -> None
+             | Some _, None -> replaced_path
+             | None, Some _ -> replaced_path'
+             | Some _, Some _ ->
+               Misc.fatal_error
+                 "Typemod.approx_with: two signature items with the same name"
+           in
+           replaced_path, item)
+        None sg
+    in
+    replaced_path,
+    List.filter_map Fun.id filtered_sg
+  and remove_from_sig_item namespace path env sg_for_env item =
+    match path, namespace, item with
+    | [s], Type, Sig_type (id, _, _, _)
+    | [s], Module, Sig_module (id, _, _, _, _)
+    | [s], Module_type, Sig_modtype (id, _, _)
+        when Ident.name id = s ->
+      Some (Pident id), None
```

Comment:
Note that in the `Type` case, this can create invalid module types by removing a ghost companion item. 
I am not sure if it can be exploited in the context of recursive modules, but this can lead to misleading error messages. For instance, in
```ocaml
module rec X: sig type t = Y.c end = X
and Y: sig class c:object end end with type c := < > = struct
   class c = object end
end
```
the current code will yield an error on a missing type `Y.c` while the root error is that the `with` constraint
```ocaml
sig class c:object end end with type c := < > 
```
is invalid, because a `with` constraint should not be able to remove the type `c` derived from the class `c` from the module type.